zio/ndjsonio/reader.go

```diff
@@ -6,28 +10,122 @@ import (
+func NewReader(reader io.Reader, zctx *resolver.Context) (*Reader, error) {
+	_, err := zctx.LookupTypeAlias("zenum", zng.TypeString)
+	if err != nil {
+		return nil, err
+	}
+
 	buffer := make([]byte, ReadSize)
+	scanner := skim.NewScanner(reader, buffer, MaxLineSize)
 	return &Reader{
-		scanner: skim.NewScanner(reader, buffer, MaxLineSize),
-		parser:  NewParser(zctx),
+		scanner: scanner,
+		stats:   ReadStats{Stats: &scanner.Stats, typeStats: &typeStats{}},
+		inf:     inferParser{zctx},
 		zctx:    zctx,
+	}, nil
+}
+
+// A Rule contains one or more matches and the name of a descriptor
+// key (in the companion Descriptors map).
+type Rule struct {
+	Name       string `json:"name"`
+	Value      string `json:"value"`
+	Descriptor string `json:"descriptor"`
+}
+
+// A TypeConfig contains a map of Descriptors, keyed by name, and a
+// list of rules defining which records should be mapped into which
+// descriptor.
+type TypeConfig struct {
+	Descriptors map[string][]interface{} `json:"descriptors"`
+	Rules       []Rule                   `json:"rules"`
+}
+
+// typeRules is used internally and is derived from TypeConfig by
+// converting its descriptors into *zng.TypeRecord s for use by the
+// ndjson typed parser.
+type typeRules struct {
+	descriptors map[string]*zng.TypeRecord
+	rules       []Rule
+}
+
+// SetTypeConfig adds a TypeConfig to the reader. Its use is optional,
+// but if used, it should be called before records are processed.  In
+// the absence of a TypeConfig, records are all parsed with the
+// inferParser. If a TypeConfig is present, records are first parsed
+// with the typeParser, and if that fails, with the inferParser.
+func (r *Reader) SetTypeConfig(tc TypeConfig) error {
+	tr := typeRules{
+		descriptors: make(map[string]*zng.TypeRecord),
+		rules:       tc.Rules,
+	}
+
+	for key, columns := range tc.Descriptors {
+		typeName, err := zjsonio.DecodeType(columns)
+		if err != nil {
+			return fmt.Errorf("error decoding type \"%s\": %s", typeName, err)
+		}
+		typ, err := r.zctx.LookupByName(typeName)
+		if err != nil {
+			return err
+		}
+		recType, ok := typ.(*zng.TypeRecord)
+		if !ok {
+			return fmt.Errorf("type not a record: \"%s\"", typeName)
+		}
+		tr.descriptors[key] = recType
+	}
+	r.typ = &typeParser{
+		zctx:          r.zctx,
+		tr:            tr,
+		stats:         r.stats.typeStats,
+		typeInfoCache: make(map[int]*typeInfo),
+	}
+	return nil
+}
+
+// Parse returns a zng.Value from the provided JSON input. The
+// function expects the input json to be an object, otherwise an error
+// is returned.
+func (r *Reader) Parse(b []byte) (zng.Value, error) {
+	val, typ, _, err := jsonparser.Get(b)
+	if err != nil {
+		return zng.Value{}, err
+	}
+	if typ != jsonparser.Object {
+		return zng.Value{}, fmt.Errorf("expected JSON type to be Object but got %s", typ)
+	}
+	if r.typ != nil {
+		if zv, err := r.typ.parseObject(val); err == nil {
+			return zv, nil
+		}
 	}
+	return r.inf.parseObject(val)
```

Comment:
Is quietly falling back to inference the right default?  If there's a typo or something in types.json, seems like it would be tricky to debug that.  Though in practice, I think its unlikely that anybody will use anything other than the default zeek types.json any time soon so this is probably fine for now.

Reply:
I don't know, and I think we'll only learn with experience. When in doubt I've leaned towards @philrz so if he thinks we should flip this around right away I'll do it. @philrz ?

Reply:
Update: after conversing with @philrz I'm going to flip this around as per @aswan 's suggestion. Not the final word, but it's what we had in boom and seems reasonable. Thanks @aswan for raising this.

Reply:
Here's additional color. Indeed, in our previous attempt at this in other software, we'd been assuming we'd want to guide the user toward a conservative bootstrapping phase. If we provide them a schema for a modern out-of-the-box Zeek environment and it works for 100% of their logs, great! If they've added customizations or additional plugins/scripts that produce additional fields/types our schema is not prepared for, we figure it's best to push them toward enduring the pain up-front to get the schema correct and "lock it in", then be notified if some later customization causes it to fail again. This is better than if some IP address or ISO timestamp gets inferred to a string and weeks later they're tearing their hair out wondering why the queries aren't matching.